File: jambeth/jambeth-merge-bytecode/src/main/java/de/osthus/ambeth/bytecode/behavior/ObjRefBehavior.java
package de.osthus.ambeth.bytecode.behavior;

import java.util.List;

import de.osthus.ambeth.bytecode.visitor.ObjRefVisitor;
import de.osthus.ambeth.ioc.annotation.Autowired;
import de.osthus.ambeth.log.ILogger;
import de.osthus.ambeth.log.LogInstance;
import de.osthus.ambeth.merge.IEntityMetaDataProvider;
import de.osthus.ambeth.merge.model.IEntityMetaData;
import de.osthus.ambeth.merge.model.IObjRef;
import de.osthus.ambeth.metadata.ObjRefEnhancementHint;
import de.osthus.ambeth.repackaged.org.objectweb.asm.ClassVisitor;

public class ObjRefBehavior extends AbstractBehavior
{
	@SuppressWarnings("unused")
	@LogInstance
	private ILogger log;

	@Autowired
	protected IEntityMetaDataProvider entityMetaDataProvider;

	@Override
	public Class<?>[] getEnhancements()
	{
		return new Class<?>[] { IObjRef.class };
	}

	@Override
	public ClassVisitor extend(ClassVisitor visitor, IBytecodeBehaviorState state, List<IBytecodeBehavior> remainingPendingBehaviors,
			List<IBytecodeBehavior> cascadePendingBehaviors)
	{
		ObjRefEnhancementHint memberHint = state.getContext(ObjRefEnhancementHint.class);
		if (memberHint == null)
		{
			return visitor;
		}
		IEntityMetaData metaData = entityMetaDataProvider.getMetaData(memberHint.getEntityType());
		visitor = new ObjRefVisitor(visitor, metaData, memberHint.getIdIndex());
		return visitor;
	}
}

File: ambeth/Ambeth.Merge.Bytecode/ambeth/bytecode/behavior/ObjRefBehavior.cs
using De.Osthus.Ambeth.Bytecode.Behavior;
using De.Osthus.Ambeth.Bytecode.Visitor;
using De.Osthus.Ambeth.Ioc.Annotation;
using De.Osthus.Ambeth.Log;
using De.Osthus.Ambeth.Merge;
using De.Osthus.Ambeth.Merge.Model;
using De.Osthus.Ambeth.Metadata;
using De.Osthus.Ambeth.Typeinfo;
using System;
using System.Collections.Generic;
using System.Text;

namespace De.Osthus.Ambeth.Bytecode.Behavior
{
public class ObjRefBehavior : AbstractBehavior
{
	[LogInstance]
	public ILogger Log { private get; set; }

	[Autowired]
	public IEntityMetaDataProvider EntityMetaDataProvider { protected get; set; }

	public override Type[] GetEnhancements()
    {
 	     return new Type[] { typeof(IObjRef) };
    }

    public override IClassVisitor Extend(IClassVisitor visitor, IBytecodeBehaviorState state, IList<IBytecodeBehavior> remainingPendingBehaviors, IList<IBytecodeBehavior> cascadePendingBehaviors)
    {
		ObjRefEnhancementHint memberHint = state.GetContext<ObjRefEnhancementHint>();
		if (memberHint == null)
		{
			return visitor;
		}
		IEntityMetaData metaData = EntityMetaDataProvider.GetMetaData(memberHint.EntityType);
		visitor = new ObjRefVisitor(visitor, metaData, memberHint.IdIndex);
		return visitor;
	}
}
}